### PR TITLE
Add the support for redis in cAdvisor's storage

### PR DESCRIPTION
**Pull request by superzhaoyy**
provide redis driver for cAdvisor's storage，we can push the data from  cAdvisor to redis.

### CODE REVIEW

**Comment by superzhaoyy** [comment]
I signed it!

File: storagedriver.go [hunk 1]
```diff
@@ -72,6 +73,18 @@ func NewMemoryStorage(backendStorageName string) (*memory.InMemoryStorage, error
 			*argDbTable,
 			*argDbName,
 		)
+	case "redis":
+		var hostname string
+		hostname, err = os.Hostname()
```
Comment (rjnagal) [inline storagedriver.go]: Hostname() won't return what you expect if cAdvisor is running inside a container. You probably want to accept an IP or hostname as a flag.

File: storage/redis/redis.go [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+func (self *redisStorage) AddStats(ref info.ContainerReference, stats *info.ContainerStats) error {
+	if stats == nil {
+		return nil
+	}
+	var seriesToFlush []byte
+	func() {
+		// AddStats will be invoked simultaneously from multiple threads and only one of them will perform a write.
+		self.lock.Lock()
+		defer self.lock.Unlock()
+		b, _ := json.Marshal(stats)
```
Comment (vmarmol) [inline storage/redis/redis.go]: let's marshal inside the if so that we only do it when we are going to write.

**Comment by vmarmol** [comment]
ok to test

**Comment by vmarmol** [comment]
this looks great @superzhaoyy, thanks for the PR!

**Comment by vmarmol** [comment]
Looks like we'll need to add the redis godep btw.

**Comment by superzhaoyy** [comment]
hi  @vmarmol     why I can't  add the redis godep   ?

**Comment by vmarmol** [comment]
Usually what works for me is this (from cAdvisor root):

```
$ godep restore
$ go get -u github.com/garyburd/redigo/redis
$ godep save ./...
```

Does that return any problems?

**Comment by superzhaoyy** [comment]
3Q@vmarmol
As you said, I successfully added redis in Godeps.

File: Godeps/Godeps.json [hunk 3]
```diff
@@ -1,6 +1,6 @@
 {
 	"ImportPath": "github.com/google/cadvisor",
-	"GoVersion": "go1.4.2",
+	"GoVersion": "go1.2.1",
```
Comment (rjnagal) [inline Godeps/Godeps.json]: Can you run with go1.4 or just change the version back in this file manually?

File: storage/redis/redis.go [hunk 4]
```diff
@@ -0,0 +1,125 @@
+// Copyright 2014 Google Inc. All Rights Reserved.
```
Comment (rjnagal) [inline storage/redis/redis.go]: 2015

File: storagedriver.go [hunk 5]
```diff
@@ -72,6 +73,22 @@ func NewMemoryStorage(backendStorageName string) (*memory.InMemoryStorage, error
 			*argDbTable,
 			*argDbName,
 		)
+	case "redis":
+	//machineName: A unique identifier to identify the host that current cAdvisor
```
Comment (rjnagal) [inline storagedriver.go]: incomplete sentence.

**Comment by rjnagal** [comment]
Looks great, just minor nits.

Can you also squash your commits. Thanks @superzhaoyy !

**Comment by superzhaoyy** [comment]
hi@rjnagal  
I have modified some mistakes that you had pointed out，but how can i squash my commits，I don't know what's the meanning .  
thank you.

**Comment by vmarmol** [comment]
@superzhaoyy that pretty much means to make your commits into less commits, so that you're "squashing" them together. This article does a good job walking you through it:

http://gitready.com/advanced/2009/02/10/squashing-commits-with-rebase.html

Let us know if you run into any trouble.

**Comment by superzhaoyy** [comment]
hi @vmarmol 
oh ,it's sad .I add more commits because i make mistakes by using git . can i give up this PR, then I give another PR ?

**Comment by vmarmol** [comment]
That works too :)

It looks like you got a rebase slightly incorrect. Git is a bit of a pain to get used to for stuff like this...

**Comment by superzhaoyy** [comment]
@vmarmol 
yes.it is sad .I am not able to skillfully use Git. I have to delete my repository,i refork google/cadvisor.
can i give up this PR, then I give another PR ? 
Thank you for your patience to answer.
:+1:

**Comment by vmarmol** [comment]
Yes you certainly can close this PR and make a new one. You shouldn't need to delete your repository, just make a new branch:

Assuming that `origin` is the upstream fork in GitHub.

```
# Grab the latest code
$ git fetch origin master

# Make a new branch based on the current code
$ git branch new-branch origin/master

# Switch to that branch
$ git checkout new-branch
```

**Comment by superzhaoyy** [comment]
thanks@vmarmol
I have make a new PR. Please close this PR.
It's nice talking with you.

**Pull request closed without merging**